### ISSUE TITLE
wellKnownSymbol during prerender | Potential upgrade issue

### ISSUE THREAD

**Issue by tctrautman** [redwoodjs/redwood#1890]
First off -- wow am I excited for pre-rendering, and thank you so much @dac09 for all the work you did to make it happen.  I've been lurking in the prerender PR and have a sense of how much work was involved -- it was exciting to see this come to fruition in the meetup presentation on Friday. 😄 

With all that in mind, I wanted to share an error I've run into while trying to prerender my home page:

```
$ yarn rw prerender --dry-run
yarn run v1.19.1
$ /.../node_modules/.bin/rw prerender --dry-run
  ✖ Prerendering / -> web/dist/index.html
    → wellKnownSymbol is not a function

Not all routes were succesfully prerendered
we could not prerender your page, but your Redwood app should still work fine.
This could mean that a library you're using does not support SSR.

TypeError: wellKnownSymbol is not a function
    at Object.<anonymous> (/.../node_modules/core-js/internals/add-to-unscopables.js:5:19)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Module._compile (/.../node_modules/pirates/lib/index.js:99:24)
    at Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Object.newLoader [as .js] (/.../node_modules/pirates/lib/index.js:104:7)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (/.../web/node_modules/core-js/modules/esnext.array.last-index.js:3:24) {
  context: [Object: null prototype] {}
}

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

As the error suggests, I've attempted to go through and check my imported libraries to see if they support SSR, but it's not always clear if they do.  As far as I can tell I've made the necessary code changes, and check `if (browser)` where needed (initializing `react-ga` and `gotrue-js`.)

So I was wondering: could we either display which libraries were triggering this error, or get a pointer to where in the page there is a problem without too much more work?  I could certainly use some help figuring this out, and I figure others might too.

This isn't an area of expertise for me, but I might have some bandwidth to attempt a PR here if I could get pointed in the right direction 😄

**Comment by dac09** [comment]
Hey @tctrautman, I've seen this before, this happens when yarn duplicates node_modules in your web/node_modules folder. To fix, try deleting these (rm -rf web/node_modules), yarn again and it should work. 

Were you by chance trying the contributor workflow from canary or upgraded React?

**Comment by tctrautman** [comment]
Thank you for the quick response!  Unfortunately it didn't quite seem to resolve my problem though 🤷‍♂️ :

#### The error before doing anything:
```
$ yarn rw prerender --dry-run
yarn run v1.19.1
$ /.../node_modules/.bin/rw prerender --dry-run
  ✖ Prerendering / -> web/dist/index.html
    → wellKnownSymbol is not a function

Not all routes were succesfully prerendered
we could not prerender your page, but your Redwood app should still work fine.
This could mean that a library you're using does not support SSR.

TypeError: wellKnownSymbol is not a function
    at Object.<anonymous> (/.../web/node_modules/core-js/internals/add-to-unscopables.js:5:19)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Module._compile (/.../node_modules/pirates/lib/index.js:99:24)
    at Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Object.newLoader [as .js] (/.../node_modules/pirates/lib/index.js:104:7)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (/.../web/node_modules/core-js/modules/esnext.array.last-index.js:3:24) {
  context: [Object: null prototype] {}
}

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

#### I deleted node_modules and reinstalled:
```
$ rm -rf web/node_modules
$ cd web/ && ls && cd ..
config                  jest.config.js          package.json            src                     yarn-error.log
dist                    jsconfig.json           public                  tailwind.config.js
$ yarn
yarn install v1.19.1
[1/5] 🔍  Validating package.json...
[2/5] 🔍  Resolving packages...
[3/5] 🚚  Fetching packages...
[4/5] 🔗  Linking dependencies...
...
[5/5] 🔨  Building fresh packages...
✨  Done in 11.60s.
$ cd web/ && ls && cd ..
config                  jest.config.js          node_modules            public                  tailwind.config.js
dist                    jsconfig.json           package.json            src                     yarn-error.log
```

#### Unfortunately seeing the same thing after reinstalling:
```

$ yarn rw prerender --dry-run
yarn run v1.19.1
$ /.../node_modules/.bin/rw prerender --dry-run
  ✖ Prerendering / -> web/dist/index.html
    → wellKnownSymbol is not a function

Not all routes were succesfully prerendered
we could not prerender your page, but your Redwood app should still work fine.
This could mean that a library you're using does not support SSR.

TypeError: wellKnownSymbol is not a function
    at Object.<anonymous> (/.../web/node_modules/core-js/internals/add-to-unscopables.js:5:19)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Module._compile (/.../node_modules/pirates/lib/index.js:99:24)
    at Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Object.newLoader [as .js] (/.../node_modules/pirates/lib/index.js:104:7)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (/.../web/node_modules/core-js/modules/esnext.array.last-index.js:3:24) {
  context: [Object: null prototype] {}
}

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

And to your questions: I wasn't trying the contributor workflow and haven't attempted to upgrade React.  I doubt it's relevant, but I _did_ have this recent irregularity on the web side due to duplicate tailwind imports: https://github.com/redwoodjs/redwood/issues/1850

**Comment by dac09** [comment]
Ah shame. Would you mind sharing your package.json and web/package.json please? 

FYI, this happens when yarn decides that the redwood packages at the root of your project don't match up to the versions in your web packages.

(A screenshot inside your web/node_modules folder would be helpful too, I expect to see a @redwoodjs folder and a corejs folder, among others which is the cause of our issue here.

**Comment by tctrautman** [comment]
Ah, thanks for the explanation.

This is sounding _a bit_ more related to the issue I linked to above ☝️ , where I was having to upgrade using tags instead of just `yarn rw upgrade`.  In it, @thedavidprice noted:

> Under the hood, there are actually two different mechanics at work. rw upgrade using a Yarn interactive to upgrade only the @redwoodjs packages... For tags, we do some custom logic to target specific packages and then basically run the set through yarn upgrade [packages@tag]

I upgraded to `0.26.2` with `yarn rw upgrade`, but maybe the fact that I upgraded to `0.25.x` with `yarn rw upgrade -t 0.25.x` confused yarn somehow?  Seems unlikely, but I don't understand the nuances of yarn.

Here are some package.json's and node_modules for you 😄 

#### package.json:
```
{
  "private": true,
  "workspaces": {
    "packages": [
      "api",
      "web"
    ]
  },
  "devDependencies": {
    "@redwoodjs/core": "0.26.2"
  },
  "eslintConfig": {
    "extends": "@redwoodjs/eslint-config"
  },
  "engines": {
    "node": ">=12",
    "yarn": ">=1.15"
  },
  "dependencies": {}
}

```

#### web/package.json:
```
{
  "name": "web",
  "version": "0.0.0",
  "private": true,
  "browserslist": {
    "development": [
      "last 1 version"
    ],
    "production": [
      "defaults",
      "not IE 11",
      "not IE_Mob 11"
    ]
  },
  "dependencies": {
    "@headlessui/react": "^0.3.1",
    "@redwoodjs/auth": "0.26.2",
    "@redwoodjs/forms": "0.26.2",
    "@redwoodjs/router": "0.26.2",
    "@redwoodjs/web": "0.26.2",
    "@stripe/stripe-js": "^1.11.0",
    "@tailwindcss/aspect-ratio": "^0.2.0",
    "@tailwindcss/forms": "^0.2.1",
    "@tailwindcss/typography": "^0.3.1",
    "aws-sdk": "^2.848.0",
    "axios": "^0.21.0",
    "gotrue-js": "^0.9.27",
    "lottie-react-web": "^2.2.2",
    "luxon": "^1.25.0",
    "prop-types": "^15.7.2",
    "react": "16.13.1",
    "react-dom": "16.13.1",
    "react-ga": "^3.3.0",
    "react-helmet": "^6.1.0",
    "react-loading-skeleton": "^2.1.1",
    "react-quill": "^1.3.5",
    "react-spinners": "^0.9.0",
    "react-trix-rte": "^1.0.7",
    "react-typist": "^2.0.5",
    "react-typist-loop": "^0.0.5"
  },
  "devDependencies": {
    "autoprefixer": "9.8.6",
    "postcss-loader": "4.0.2",
    "tailwindcss": "npm:@tailwindcss/postcss7-compat"
  }
}

```

#### And here's my web/node_modules folder:
![Screen Shot 2021-02-28 at 9 07 47 PM](https://user-images.githubusercontent.com/4513085/109454435-15903a00-7a09-11eb-9a1f-681af48acb47.png)

**Comment by dac09** [comment]
It definitely looks like something has gone wrong in the yarn resolution. Mind trying the following?

1. Remove yarn.lock file `rm yarn.lock`
2. Remove installed node_modules and sub node_modules `rm node_modules && rm web/node_modules && rm api/node_modules`
3. Run installation again `yarn install`

What we're hoping to see is web/node_modules should _not_ contain `@redwoodjs` and `core-js`

You're right, it could possibly be related to upgrade, and maybe something went wrong during the process.

**Comment by tctrautman** [comment]
Oh, no worries at all -- I'm honestly stunned at how responsive you've been over the weekend.  I really appreciate it.

I just ran through the process of deleting yarn.lock and the three node_modules folders, then reinstalling.  I'm seeing that it updated my `yarn.lock` file, but I'm unfortunately still seeing `@redwoodjs` and `core-js` in `web/node_modules` and get the same error when running `yarn rw prerender --dry-run`:

![Screen Shot 2021-03-01 at 8 24 39 AM](https://user-images.githubusercontent.com/4513085/109527422-7bf87500-7a68-11eb-8db4-9740e7925e0d.png)

**Comment by dac09** [comment]
Let me look into this further @tctrautman, thank you for the great bug report :)

**Comment by dac09** [comment]
I tried to repro this in **_so many ways_**, and just couldn't..... 

![mild_panic](https://user-images.githubusercontent.com/1521877/109557386-f3adba80-7acf-11eb-86e5-8533cb377909.png)


UNTIL I copied your package.json-s line for line. I realised your `react` and `react-dom` versions are pinned to 16.3.1 in web/package.json. It should be like this:

```diff
-    "react": "16.13.1",
-   "react-dom": "16.13.1",
+   "react": "^16.13.1",
+   "react-dom": "^16.13.1",
```
It's missing the `^` from the version. 

Don't you love it when the little things catch you out? Although you were seeing this issue with Prerender, it can cause all sorts of havoc sometimes,  it did look like it was duplicating way more than just redwood and react modules. Glad you posted this issue.

Please let me know if this solves your issue!

**Comment by tctrautman** [comment]
Wooo!  Thank you @dac09 -- that did the trick.  And wow, good eye.  I really appreciate you putting so much time into looking into this.

![thank you](https://user-images.githubusercontent.com/4513085/109585566-98bc9900-7ab8-11eb-9d3c-bc81c8e173ec.gif)

I was thinking about how this happened... I think this _might_ be an artifact from the workaround mentioned in the [0.23 release](https://community.redwoodjs.com/t/redwoodjs-v0-23/1693):

> The fix --> downgrade both “react” and “react-dom” in web/package.json to 16.13.1

Maybe this wasn't how I was supposed to downgrade, but I believe that's how I ended up here.  I'm closing the ticket, but maybe we should add a note in the 0.26.x release notes to look out for this error if this is how you handled upgrading 0.23?

Thanks again!

**Comment by dac09** [comment]
100 points to Hufflepuff! You're very welcome, glad it's sorted, thanks for bearing with! 

**tldr;** 
Pinning specific versions of react and react-dom in web/package.json can lead to odd issues in how yarn workspaces resolves modules.
